src/r1cs/mod.rs

```diff
@@ -510,6 +659,70 @@ impl<E: Engine> RelaxedR1CSInstance<E> {
       u,
     }
   }
+
+  /// Folds an incoming `RelaxedR1CSInstance` into the current one
+  pub fn fold_relaxed(
+    &self,
+    U2: &RelaxedR1CSInstance<E>,
+    comm_T: &Commitment<E>,
+    r: &E::Scalar,
+  ) -> RelaxedR1CSInstance<E> {
+    let (X1, u1, comm_W_1, comm_E_1) =
+      (&self.X, &self.u, &self.comm_W.clone(), &self.comm_E.clone());
+    let (X2, u2, comm_W_2, comm_E_2) = (&U2.X, &U2.u, &U2.comm_W, &U2.comm_E);
+
+    // weighted sum of X, comm_W, comm_E, and u
+    let X = X1
+      .par_iter()
+      .zip(X2)
+      .map(|(a, b)| *a + *r * *b)
+      .collect::<Vec<E::Scalar>>();
+    let comm_W = *comm_W_1 + *comm_W_2 * *r;
+    let comm_E = *comm_E_1 + *comm_T * *r + *comm_E_2 * *r * *r;
+    let u = *u1 + *r * *u2;
+
+    RelaxedR1CSInstance {
+      comm_W,
+      comm_E,
+      X,
+      u,
+    }
+  }
+
+  pub fn derandomize_commits(
```

Comment:
Nit: I think we should have a method called `derandomize` in `RelaxedR1CSWitness` and `RelaxedR1CSInstance` rather than have then separately, isn't it?

The prover can call derandomize on witness types first to get back blinds. Then it can pass those blinds to derandomize method on the instances to get back derandomized instances.

This will make the code far cleaner and move these methods to the right types. Isn't it?

Reply:
This sounds more elegant, will do.